### PR TITLE
Monitoring Tool for rendering engine

### PR DESCRIPTION
**Pull request by samcake**
This work exposes a bunch of counters from the gpu::Context and the Engine layers. THese are collected once every frame in the "EngineStats" job which is automatically added to the Engine.
THen these counters are directly exposed to JS api as memeber of the Render.Stats object.

THis PR also add a monitoring script that spawn a ui displaying these values live.

To test the stats, run the script "utilities/tools/render/renderStats.js" and try understanding the values reported. 
feedback welcome on the names!

### CODE REVIEW

File: interface/src/Application.cpp [hunk 1]
```diff
@@ -269,7 +269,7 @@ class DeadlockWatchdogThread : public QThread {
     void run() override {
         while (!_quit) {
             QThread::sleep(HEARTBEAT_UPDATE_INTERVAL_SECS);
-#ifdef NDEBUG
+#ifdef ____NDEBUG
```
Comment (ZappoMan) [inline interface/src/Application.cpp]: what's this?

**Comment by samcake** [comment]
I know some code exposing the counter atomics in Texture.h and Resource.h is not very clean yet, i will take care of that, wanted to share what i have here before too long

File: examples/utilities/tools/render/stats.qml [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+        function evalEvenHeight() {
+            // Why do we have to do that manually ? cannot seem to find a qml / anchor / layout mode that does that ?
+            return (height - spacing * (children.length - 1)) / children.length
+        }
+
+        PlotPerf {
+            title: "Num Buffers"
+            config: stats.config
+            height: parent.evalEvenHeight()
+            parameters: "1::0:CPU-numBuffers-#00B4EF:GPU-numGPUBuffers-#1AC567"
```
Comment (jherico) [inline examples/utilities/tools/render/stats.qml]: why all the encoding and parsing instead of just setting the values directly on each PlotPerf?

File: interface/src/Application.cpp (hunk 1 above)
Comment (jherico) [inline interface/src/Application.cpp]: ???

**Comment by hifi-gustavo** [comment]
Build console output for commit: c23640249e883713e021d222013b2600ea71a75d available at: https://gist.github.com/68f1de1488c5c55e451a<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/HighFidelity-Alpha-PR7481.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/HighFidelity-Alpha-PR7481.dmg

File: libraries/gpu/src/gpu/Resource.h [hunk 3]
```diff
@@ -109,7 +110,21 @@ class Resource {
 };
 
 class Buffer : public Resource {
+    static std::atomic<int> _numBuffers;
```
Comment (jherico) [inline libraries/gpu/src/gpu/Resource.h]: you can't have negative numbers of buffers, so this should be unsigned, either `size_t` or `Size`

File: examples/utilities/tools/render/stats.qml (hunk 2 above)
Comment (samcake) [inline examples/utilities/tools/render/stats.qml]: I guess i started with a REpeater and that s why i ended up with that style
yep....

File: libraries/gpu/src/gpu/Resource.h (hunk 3 above)
Comment (samcake) [inline libraries/gpu/src/gpu/Resource.h]: yep, i realize that unix complained about that too

**Comment by hifi-gustavo** [comment]
Build console output for commit: 5190878db21b28047a81b14f77c9528b002ca1cf available at: https://gist.github.com/68f1de1488c5c55e451a<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/HighFidelity-Alpha-PR7481.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/HighFidelity-Alpha-PR7481.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: 73613040841858e85e1b95abab03f3c285c62d89 available at: https://gist.github.com/68f1de1488c5c55e451a<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/HighFidelity-Alpha-PR7481.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/HighFidelity-Alpha-PR7481.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: c96d5b093cd88e2493224de4535127f3aeb0d162 available at: https://gist.github.com/68f1de1488c5c55e451a<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/HighFidelity-Alpha-PR7481.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/HighFidelity-Alpha-PR7481.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: c182330b6ee469fe7a1b71211c3f2ac07ae0f2ba available at: https://gist.github.com/68f1de1488c5c55e451a<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/HighFidelity-Alpha-PR7481.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/HighFidelity-Alpha-PR7481.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: 5a3896b1de88d5934c12171f2e568031a6ebe055 available at: https://gist.github.com/68f1de1488c5c55e451a<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/HighFidelity-Alpha-PR7481.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/HighFidelity-Alpha-PR7481.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: 7402dafc91a98464cb1dc330b3236e871be63598 available at: https://gist.github.com/68f1de1488c5c55e451a<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/HighFidelity-Alpha-PR7481.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/HighFidelity-Alpha-PR7481.dmg

**Comment by cozza13** [comment]
how do you test this?

**Comment by samcake** [comment]
@chris, By running the script :"utilities/tools/render/renderStats.js" 
You should see graphs representing the current memory consumption for Buffers and Textures both on CPU and GPU memory.
THe goal of this PR is to expose these information. 
I validated that they were correct.
QA would make sure that looking at different scenes, the values reported are making sense.
For example by comparing the memory values reported with what you would get from the Performance Monitor fo Windows that would tell exactly how much memory is allocated in TOTAL for the interface.exe process.
You would need to validate that the values reported by the official tool from windows and our numbers are coherent.

**Comment by Menithal** [comment]
Seems to be updating along pretty decently.



As a side note, I suggest making the values much brighter.

**Comment by Menithal** [comment]
@samcake  as a side note the values given are for the non-gpu memory in the performance monitoring, it wouldnt give info on the gpu side.

instead GPU-z reports much higher memory use however than what High Fidelity does.


Offset by 600 (base line if same windows were open but high fidelity, then add to it high fidelity. 1017.2 Mb for gpu memmory, memmory load according to GPU (without baseline) 1900 mb. So 900 mb is not accounted for.

File: libraries/gpu/src/gpu/Texture.cpp [hunk 4]
```diff
@@ -59,15 +118,15 @@ const Texture::PixelsPointer Texture::Storage::getMipFace(uint16 level, uint8 fa
 
 void Texture::Storage::notifyMipFaceGPULoaded(uint16 level, uint8 face) const {
     PixelsPointer mipFace = getMipFace(level, face);
-    if (mipFace && (_type != TEX_CUBE)) {
-        mipFace->_isGPULoaded = true;
-        mipFace->_sysmem.resize(0);
+   // if (mipFace && (_type != TEX_CUBE)) {
```
Comment (ZappoMan) [inline libraries/gpu/src/gpu/Texture.cpp]: dead code? why did you change this logic?

File: libraries/gpu/src/gpu/Texture.h [hunk 5]
```diff
@@ -138,7 +138,19 @@ class Sampler {
 };
 
 class Texture : public Resource {
+    static std::atomic<uint32_t> _textureCPUCount;
+    static std::atomic<Size> _textureCPUMemoryUsage;
+    static void updateTextureCPUMemoryUsage(Size prevObjectSize, Size newObjectSize);
+
+public:
+    static std::atomic<uint32_t> _textureGPUCount;
```
Comment (ZappoMan) [inline libraries/gpu/src/gpu/Texture.h]: coding standard: I keep telling you this... the coding standard says "_" is for PRIVATE members... and public members should not have a prefix.

File: libraries/render/src/render/EngineStats.h [hunk 6]
```diff
@@ -0,0 +1,91 @@
...
+    };
+
+    class EngineStats {
+    public:
+        using Config = EngineStatsConfig;
+        using JobModel = Job::Model<EngineStats, Config>;
+
+        EngineStats() { _frameTimer.start(); }
+
+        gpu::ContextStats _gpuStats;
```
Comment (ZappoMan) [inline libraries/render/src/render/EngineStats.h]: coding standard: "_" for PRIVATE not PUBLIC

File: libraries/render/src/render/EngineStats.h [hunk 7]
```diff
@@ -0,0 +1,91 @@
...
+
+        void configure(const Config& configuration) {}
+        void run(const SceneContextPointer& sceneContext, const RenderContextPointer& renderContext);
+
+        QElapsedTimer _frameTimer;
```
Comment (ZappoMan) [inline libraries/render/src/render/EngineStats.h]: coding standard: "_" for PRIVATE not PUBLIC

**Comment by samcake** [comment]
Great feedback @Menithal !
I wil definitely try to find these missing memory.
We are not counting everything, in particular we don;t count anything allocated by the DisplayPLugin (final swapbuffer and potentially extra textures) 
I ll try to add them to the count so we have a more accurate data
Thanks

**Comment by hifi-gustavo** [comment]
Build console output for commit: b5028acde1a878098ab5eba8962af2ad878a51be available at: https://gist.github.com/68f1de1488c5c55e451a<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/HighFidelity-Alpha-PR7481.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/HighFidelity-Alpha-PR7481.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: aa085f69559e020e82632bd68918abae9a6579c0 available at: https://gist.github.com/68f1de1488c5c55e451a<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/HighFidelity-Alpha-PR7481.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/HighFidelity-Alpha-PR7481.dmg
